lib/rails_semantic_logger/engine.rb

```diff
@@ -108,7 +108,13 @@ class Engine < ::Rails::Engine
       Resque.logger           = SemanticLogger[Resque] if defined?(Resque) && Resque.respond_to?(:logger)
 
       # Replace the Sidekiq logger
-      Sidekiq::Logging.logger = SemanticLogger[Sidekiq] if defined?(Sidekiq)
+      if defined?(Sidekiq)
+        if defined?(Sidekiq::Logging)
+          Sidekiq::Logging.logger = SemanticLogger[Sidekiq]
+        else
+          Sidekiq.logger = SemanticLogger[Sidekiq]
```

Comment:
Is this something that would help this PR?

```suggestion
          Sidekiq.logger = SemanticLogger[Sidekiq]
          Sidekiq.logger.extend(Sidekiq::LogContext)
```

Reply:
@bobvanderlinden sorry for the delay. Is there a branch I can use to test this?

Reply:
I tried to patch my rails_semantic_logger with this snippet, but I'm seeing this error

```
sidekiq_1_497f65c0a128 | uninitialized constant Sidekiq::LogContext
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:79:in `block in load_missing_constant'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:8:in `without_bootsnap_cache'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:79:in `rescue in load_missing_constant'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/active_support.rb:58:in `load_missing_constant'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/rails_semantic_logger-4.4.2/lib/rails_semantic_logger/engine.rb:118:in `block in <class:Engine>'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/lazy_load_hooks.rb:69:in `block in execute_hook'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/lazy_load_hooks.rb:62:in `with_execution_control'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/lazy_load_hooks.rb:67:in `execute_hook'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/lazy_load_hooks.rb:52:in `block in run_load_hooks'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/lazy_load_hooks.rb:51:in `each'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/activesupport-5.2.3/lib/active_support/lazy_load_hooks.rb:51:in `run_load_hooks'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/railties-5.2.3/lib/rails/application/bootstrap.rb:81:in `block in <module:Bootstrap>'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/railties-5.2.3/lib/rails/initializable.rb:32:in `instance_exec'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/railties-5.2.3/lib/rails/initializable.rb:32:in `run'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/railties-5.2.3/lib/rails/initializable.rb:61:in `block in run_initializers'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/2.5.0/tsort.rb:228:in `block in tsort_each'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/2.5.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/2.5.0/tsort.rb:431:in `each_strongly_connected_component_from'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/2.5.0/tsort.rb:349:in `block in each_strongly_connected_component'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/2.5.0/tsort.rb:347:in `each'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/2.5.0/tsort.rb:347:in `call'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/2.5.0/tsort.rb:347:in `each_strongly_connected_component'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/2.5.0/tsort.rb:226:in `tsort_each'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/2.5.0/tsort.rb:205:in `tsort_each'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/railties-5.2.3/lib/rails/initializable.rb:60:in `run_initializers'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/railties-5.2.3/lib/rails/application.rb:361:in `initialize!'
sidekiq_1_497f65c0a128 | /app/config/environment.rb:5:in `<top (required)>'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/sidekiq-6.0.0/lib/sidekiq/cli.rb:248:in `require'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/sidekiq-6.0.0/lib/sidekiq/cli.rb:248:in `boot_system'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/sidekiq-6.0.0/lib/sidekiq/cli.rb:37:in `run'
sidekiq_1_497f65c0a128 | /usr/local/bundle/gems/sidekiq-6.0.0/bin/sidekiq:12:in `<top (required)>'
sidekiq_1_497f65c0a128 | /usr/local/bundle/bin/sidekiq:23:in `load'
sidekiq_1_497f65c0a128 | /usr/local/bundle/bin/sidekiq:23:in `<top (required)>'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/lib/bundler/cli/exec.rb:74:in `load'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/lib/bundler/cli/exec.rb:74:in `kernel_load'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/lib/bundler/cli/exec.rb:28:in `run'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/lib/bundler/cli.rb:463:in `exec'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/lib/bundler/cli.rb:27:in `dispatch'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/lib/bundler/cli.rb:18:in `start'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/exe/bundle:30:in `block in <top (required)>'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'
sidekiq_1_497f65c0a128 | /usr/local/lib/ruby/gems/2.5.0/gems/bundler-1.17.2/exe/bundle:22:in `<top (required)>'
sidekiq_1_497f65c0a128 | /usr/local/bin/bundle:23:in `load'
sidekiq_1_497f65c0a128 | /usr/local/bin/bundle:23:in `<main>'
```

Reply:
This line instead (taken from [here](https://github.com/mperham/sidekiq/issues/4239#issuecomment-528801658)) seems not to cause issues however, although I have to admit, I'm not entirely sure what either of those lines do

```ruby
Sidekiq.logger.class.alias_method(:with_context, :tagged)
```

Reply:
Yes. Creating the with_context is a bit dirtier imo, but if Sidekiq::LogContext is indeed not exposed, it is the only way at the moment it seems :(.

Reply:
I think your original PR however works together with sidekiq master though (when 6.0.1 is released). see https://github.com/mperham/sidekiq/issues/4254

So this extra option won't be necessary then, right?

Reply:
`define_singleton_method` also seems to alter the underlying class. I've made a separate branch that properly implements the `with_context` in a Sidekiq-specific logger class. I think this is the cleanest solution for now, but Sidekiq still overrides `with_context` with the current master version.

Reply:
@bobvanderlinden I'm not sure I follow you precisely. But even if I completely agree and this is a cleaner/nicer/better way, then shouldn't this go into Sidekiq? (where the ugly/dirty/sub-optimal solution lives). At least make a case to change it inside Sidekiq and hopefully other logging libraries would benefit as well?

Ideally Sidekiq plays nicely with any loggers, and the Semantic Logger requires minimal changes to plug-in.

What am I missing?

Reply:
Sidekiq 'wants' to use a `with_context` function that is in no-way supported by the Rails standard loggers. There is no standard interface for such a function. SemanticLogger has chosen to use `tagged` as the name for such a function; Sidekiq has chosen to use `with_context` as the name. Both do the same thing. Forcing Sidekiq to use `tagged` will not make sense as then Sidekiq is adhering to the interface chosen by `SemanticLogger` (which is likely different for other logging frameworks).

So, Sidekiq always wants to have the ability to enrich its log lines with extra information and does so using the `with_context` function. It forces any logger assigned to Sidekiq to have such a function.

For convenience (and this is where it is in the wrong imo) it creates the `with_context` function to the logger class if it doesn't exist. This is bad.

We can support that quite easily by wrapping our log function with one that aliases `with_context` to `tagged`. We only need to use such a wrapper for Sidekiq specifically and do not need to change any generic SemanticLogger functionality.

It would be nice if Sidekiq didn't use the `with_context` function at all, but I can very well imagine that enriching log lines with extra information is something we'd also want them to do.

Reply:
Thanks for explaining, @bobvanderlinden. Wouldn't it be more fruitful to try to give this feedback to the Sidekiq project? (in case you haven't already, apologies if you have).

Even if we don't agree with what Sidekiq is doing, it is the de-facto standard library for async jobs in rails, and SemanticLogger should play nicely with it right? At least if we want SemanticLoggerRails to support Sidekiq out of the box...

Besides that, the last paragraph confused me a bit... could you clarify what you mean?

> It would be nice if Sidekiq didn't use the with_context function at all, but I can very well imagine that enriching log lines with extra information is something we'd also want them to do.

Reply:
I have submitted the following issue at Sidekiq: https://github.com/mperham/sidekiq/issues/4296

The last paragraph is about Sidekiq using with_context, which is a function not part of any Rails loggers. Sidekiq using that function means it is relying on non-standard functionality. It would be nice if Sidekiq didn't rely on non-standard functionality, but I can see why they want such a feature: to enrich their logs with extra information.

The issue mentioned might give some more insight as well.

Reply:
Thanks again, @bobvanderlinden. I missed this sidekiq issue. Looks like it's going forward well and you make a good case there. I'll follow both issues. Let me know if there's anything I can help with. Thanks for doing this!